larpandoracontent/LArTwoDReco/TrackInEMShowerAlgorithm.cc

```diff
@@ -0,0 +1,892 @@
+
+    return (lhsProjectedDistance < rhsProjectedDistance);
+}
+ 
+//------------------------------------------------------------------------------------------------------------------------------------------
+//------------------------------------------------------------------------------------------------------------------------------------------
+    
+TrackInEMShowerAlgorithm::TrackInEMShowerAlgorithm() :
+    m_minCaloHits(25), 
+    m_minSeparationDistance(27.f),
+    m_maxPredictedMergePointOffset(5.f),
+    m_slidingFitWindow(20),
+    m_mergePointMinCosAngleDeviation(0.9995),
+    m_minClusterLengthSum(75.f),
+    m_minDirectionDeviationCosAngle(0.99),
+    m_distanceFromLine(0.35),
+    m_maxTrackGaps(2),
+    m_lineSegmentLength(7.f),
+    m_maxHitDistanceFromCluster(4.f),
+    m_maxHitSeparationForConnectedCluster(4.f),
+    m_maxDistanceFromMainTrack(0.75)
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+StatusCode TrackInEMShowerAlgorithm::Run()
+{
+    const ClusterList *pClusterList = NULL;
+    PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::GetCurrentList(*this, pClusterList));
+
+    const CaloHitList *pCaloHitList = NULL;
+    PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::GetCurrentList(*this, pCaloHitList));
+    
+    ClusterVector clusterVector;
+    this->SelectCleanClusters(pClusterList, clusterVector);
+
+    TwoDSlidingFitResultMap microSlidingFitResultMap;
+    TwoDSlidingFitResultMap macroSlidingFitResultMap;
+    this->InitialiseSlidingFitResultMaps(clusterVector, microSlidingFitResultMap, macroSlidingFitResultMap);
+
+    // Find longest length association and conitnue merging until no further merges can be made
+    bool mergeMade(false);
+    do
+    {
+        ClusterAssociation clusterAssociation;
+        if(!this->FindBestClusterAssociation(clusterVector, microSlidingFitResultMap, macroSlidingFitResultMap, clusterAssociation))
+            break;
+        
+        ClusterToCaloHitListMap clusterToCaloHitListMap;
+        CaloHitVector extrapolatedCaloHitVector;
+        this->GetExtrapolatedCaloHits(clusterAssociation, pClusterList, extrapolatedCaloHitVector, clusterToCaloHitListMap);
+        
+        if (extrapolatedCaloHitVector.empty())
+            break;
+        
+        if (!this->IsTrackContinuous(clusterAssociation, extrapolatedCaloHitVector))
+            break;
+        
+        this->RefineTracks(clusterAssociation, microSlidingFitResultMap, macroSlidingFitResultMap, clusterVector, extrapolatedCaloHitVector);
+        
+        this->MergeHits(clusterAssociation, clusterToCaloHitListMap, microSlidingFitResultMap, macroSlidingFitResultMap, clusterVector);
+
+        this->UpdateSlidingFitResultMap(clusterVector, microSlidingFitResultMap, macroSlidingFitResultMap);
+
+        mergeMade = true;
```

Comment:
This would actually be super beneficial! It would stop merging of tracks in the shower after the main muon track has been merged.

Reply:
Although perhaps we can safeguard even further by using the main track cluster we had created in the previous loop iteration (if there was one)

Reply:
Actually this can already be done! 

So essentially the cluster vector contains the clusters that the algorithm considers when searching for a CR pair. So, in each passing of the merging process in the cleanup stage I remove clusters from the map and cluster vector that are going to be changed (so the main track clusters and all small shower clusters that have hits that belong to the main track). Then when I update for the next iteration I could perhaps only add back in the main track cluster and so in the future only this cluster and others not considered in the previous merge will be considered. But I think i'll test this with metrics